## Homework 6 Data Cleaning 

###  DatasET  : NYPD Motor Vehicle Collisions
This is a breakdown of every collision in NYC by location and injury. This data is collected because the NYC Council passed Local Law #11 in 2011. This data is manually run every month and reviewed by the TrafficStat Unit before being posted on the NYPD website. Each record represents a collision in NYC by city, borough, precinct and cross street. This data can be used by the public to see how dangerous/safe intersections are in NYC. The information is presented in pdf and excel format to allow the casual user to just view the information in the easy to read pdf format or use the excel files to do a more in-depth analysis.

### Dataset link : https://data.cityofnewyork.us/Public-Safety/NYPD-Motor-Vehicle-Collisions/h9gi-nx95/data


### Goals : To clean the dataset by eliminating NaN values , droping NA columns , parsing Dates,etc

##  Importing Dataset

In [4]:
import pandas as pd

Importing the CSV file of the dataset and storing it in dataframe named "data"

In [15]:
data = pd.read_csv("C:/Users/akhil/Desktop/MITA SEM 1/Intro to Data Science/Homework/Data Cleaning/NYPD_Motor_Vehicle_Collisions.csv")

C:\Users\akhil\Anaconda3\lib\site-packages\IPython\core\interactiveshell.py:2785: DtypeWarning: Columns (3) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [17]:
print(data.shape) # size of dataset

(1360317, 29)


In [19]:
data.head() # data cols and rows

DATE  TIME    BOROUGH ZIP CODE   LATITUDE  LONGITUDE  \
0  10/08/2018  0:00   BROOKLYN    11211  40.713074 -73.952095   
1  10/08/2018  0:00   BROOKLYN    11224  40.577908 -74.008180   
2  10/08/2018  0:00  MANHATTAN    10036  40.758680 -73.983360   
3  10/08/2018  0:00        NaN      NaN  40.708324 -73.843140   
4  10/08/2018  0:00        NaN      NaN  40.763428 -73.965220   

                  LOCATION                    ON STREET NAME  \
0  (40.713074, -73.952095)                               NaN   
1   (40.577908, -74.00818)                               NaN   
2    (40.75868, -73.98336)                               NaN   
3   (40.708324, -73.84314)  JACKIE ROBINSON PKWY               
4   (40.763428, -73.96522)                               NaN   

  CROSS STREET NAME                           OFF STREET NAME  \
0               NaN  471       KEAP STREET                      
1               NaN  4212      HIGHLAND AVENUE                  
2               NaN  157       WEST 47 STREET                   
3               NaN                                       NaN   
4               NaN  3 AVENUE                                   

          ...           CONTRIBUTING FACTOR VEHICLE 2  \
0         ...                             Unspecified   
1         ...          Driver Inattention/Distraction   
2         ...                             Unspecified   
3         ...                             Unspecified   
4         ...                             Unspecified   

   CONTRIBUTING FACTOR VEHICLE 3  CONTRIBUTING FACTOR VEHICLE 4  \
0                            NaN                            NaN   
1                            NaN                            NaN   
2                            NaN                            NaN   
3                            NaN                            NaN   
4                            NaN                            NaN   

   CONTRIBUTING FACTOR VEHICLE 5  UNIQUE KEY  \
0                            NaN     3996168   
1                            NaN     3996787   
2                            NaN     3997603   
3                            NaN     3996194   
4                            NaN     3995663   

                   VEHICLE TYPE CODE 1                  VEHICLE TYPE CODE 2  \
0                                Sedan  Station Wagon/Sport Utility Vehicle   
1  Station Wagon/Sport Utility Vehicle  Station Wagon/Sport Utility Vehicle   
2                                Sedan                                 Taxi   
3                                Sedan                                Sedan   
4                                 Taxi                        Pick-up Truck   

   VEHICLE TYPE CODE 3 VEHICLE TYPE CODE 4 VEHICLE TYPE CODE 5  
0                  NaN                 NaN                 NaN  
1                  NaN                 NaN                 NaN  
2                  NaN                 NaN                 NaN  
3                  NaN                 NaN                 NaN  
4                  NaN                 NaN                 NaN  

[5 rows x 29 columns]

In [21]:
data.index

RangeIndex(start=0, stop=1360317, step=1)

In [24]:
data.columns # Names of the Columns

Index(['DATE', 'TIME', 'BOROUGH', 'ZIP CODE', 'LATITUDE', 'LONGITUDE',
       'LOCATION', 'ON STREET NAME', 'CROSS STREET NAME', 'OFF STREET NAME',
       'NUMBER OF PERSONS INJURED', 'NUMBER OF PERSONS KILLED',
       'NUMBER OF PEDESTRIANS INJURED', 'NUMBER OF PEDESTRIANS KILLED',
       'NUMBER OF CYCLIST INJURED', 'NUMBER OF CYCLIST KILLED',
       'NUMBER OF MOTORIST INJURED', 'NUMBER OF MOTORIST KILLED',
       'CONTRIBUTING FACTOR VEHICLE 1', 'CONTRIBUTING FACTOR VEHICLE 2',
       'CONTRIBUTING FACTOR VEHICLE 3', 'CONTRIBUTING FACTOR VEHICLE 4',
       'CONTRIBUTING FACTOR VEHICLE 5', 'UNIQUE KEY', 'VEHICLE TYPE CODE 1',
       'VEHICLE TYPE CODE 2', 'VEHICLE TYPE CODE 3', 'VEHICLE TYPE CODE 4',
       'VEHICLE TYPE CODE 5'],
      dtype='object')

### Retriving all the null values from the data

In [26]:
print(data.isnull().head()) 

    DATE   TIME  BOROUGH  ZIP CODE  LATITUDE  LONGITUDE  LOCATION  \
0  False  False    False     False     False      False     False   
1  False  False    False     False     False      False     False   
2  False  False    False     False     False      False     False   
3  False  False     True      True     False      False     False   
4  False  False     True      True     False      False     False   

   ON STREET NAME  CROSS STREET NAME  OFF STREET NAME         ...           \
0            True               True            False         ...            
1            True               True            False         ...            
2            True               True            False         ...            
3           False               True             True         ...            
4            True               True            False         ...            

   CONTRIBUTING FACTOR VEHICLE 2  CONTRIBUTING FACTOR VEHICLE 3  \
0                          False                 

In [89]:
data

DATE   TIME        BOROUGH ZIP CODE   LATITUDE  LONGITUDE  \
0        10/08/2018   0:00       BROOKLYN    11211  40.713074 -73.952095   
1        10/08/2018   0:00       BROOKLYN    11224  40.577908 -74.008180   
2        10/08/2018   0:00      MANHATTAN    10036  40.758680 -73.983360   
3        10/08/2018   0:00            NaN      NaN  40.708324 -73.843140   
4        10/08/2018   0:00            NaN      NaN  40.763428 -73.965220   
5        10/08/2018   0:00         QUEENS    11435  40.700768 -73.810190   
6        10/08/2018   0:01            NaN      NaN  40.704712 -73.727425   
7        10/08/2018   0:10      MANHATTAN    10013  40.724564 -74.007790   
8        10/08/2018   0:15          BRONX    10469  40.871810 -73.855560   
9        10/08/2018   0:17            NaN      NaN        NaN        NaN   
10       10/08/2018   0:17            NaN      NaN  40.706944 -73.917740   
11       10/08/2018   0:25      MANHATTAN    10006  40.710114 -74.012665   
12       10/08/2018   0:30          BRONX    10472  40.831287 -73.881730   
13       10/08/2018   0:30      MANHATTAN    10027  40.807960 -73.949080   
14       10/08/2018   0:30            NaN      NaN  40.716030 -73.816370   
15       10/08/2018   0:44       BROOKLYN    11201  40.691586 -73.999270   
16       10/08/2018   0:44            NaN      NaN  40.753513 -73.988790   
17       10/08/2018   0:45         QUEENS    11355  40.740173 -73.811264   
18       10/08/2018  10:00            NaN      NaN  40.766434 -73.837670   
19       10/08/2018  10:00            NaN      NaN  40.592846 -73.796160   
20       10/08/2018  10:00            NaN      NaN  40.716957 -73.826550   
21       10/08/2018  10:15            NaN      NaN  40.764587 -73.904520   
22       10/08/2018  10:15            NaN      NaN  40.860600 -73.911020   
23       10/08/2018  10:16      MANHATTAN    10029  40.788580 -73.955284   
24       10/08/2018  10:25          BRONX    10473  40.822086 -73.851906   
25       10/08/2018  10:29            NaN      NaN  40.663315 -73.989770   
26       10/08/2018  10:29            NaN      NaN  40.859850 -73.935520   
27       10/08/2018  10:30       BROOKLYN    11204  40.633500 -73.980520   
28       10/08/2018  10:30         QUEENS    11373  40.740402 -73.887100   
29       10/08/2018  10:30         QUEENS    11418  40.699290 -73.831540   
...             ...    ...            ...      ...        ...        ...   
1360287  07/01/2012   8:27       BROOKLYN    11216  40.680857 -73.938546   
1360288  07/01/2012   8:27            NaN      NaN  40.708797 -73.727620   
1360289  07/01/2012   8:30            NaN      NaN        NaN        NaN   
1360290  07/01/2012   8:34       BROOKLYN    11213  40.668798 -73.931120   
1360291  07/01/2012   8:37         QUEENS    11435  40.694545 -73.802112   
1360292  07/01/2012   8:43            NaN      NaN        NaN        NaN   
1360293  07/01/2012   8:58          BRONX    10466  40.891840 -73.862740   
1360294  07/01/2012   9:00         QUEENS    11101  40.744131 -73.927047   
1360295  07/01/2012   9:00      MANHATTAN    10022  40.757859 -73.975592   
1360296  07/01/2012   9:15       BROOKLYN    11233  40.675380 -73.908196   
1360297  07/01/2012   9:15         QUEENS    11421  40.696308 -73.851549   
1360298  07/01/2012   9:15         QUEENS    11429  40.711860 -73.735538   
1360299  07/01/2012   9:15  STATEN ISLAND    10310  40.634748 -74.112253   
1360300  07/01/2012   9:20            NaN      NaN        NaN        NaN   
1360301  07/01/2012   9:24         QUEENS    11411  40.690917 -73.728179   
1360302  07/01/2012   9:30          BRONX    10475  40.861637 -73.824326   
1360303  07/01/2012   9:30       BROOKLYN    11219  40.626826 -73.996790   
1360304  07/01/2012   9:30         QUEENS    11368  40.755336 -73.843245   
1360305  07/01/2012   9:30       BROOKLYN    11207  40.678163 -73.897477   
1360306  07/01/2012   9:39       BROOKLYN    11226  40.646518 -73.958105   
1360307  07/01/2012   9:40          BRONX    10462  40.839882 -

**Looking into the NULL values data in df getting the SUM of these null values with each column**

In [90]:
print(data.isnull().sum())

DATE                                   0
TIME                                   0
BOROUGH                           397642
ZIP CODE                          397766
LATITUDE                          224304
LONGITUDE                         224304
LOCATION                          224304
ON STREET NAME                    273211
CROSS STREET NAME                 375723
OFF STREET NAME                  1135465
NUMBER OF PERSONS INJURED              0
NUMBER OF PERSONS KILLED               0
NUMBER OF PEDESTRIANS INJURED          0
NUMBER OF PEDESTRIANS KILLED           0
NUMBER OF CYCLIST INJURED              0
NUMBER OF CYCLIST KILLED               0
NUMBER OF MOTORIST INJURED             0
NUMBER OF MOTORIST KILLED              0
CONTRIBUTING FACTOR VEHICLE 1       7315
CONTRIBUTING FACTOR VEHICLE 2     196019
CONTRIBUTING FACTOR VEHICLE 3    1272670
CONTRIBUTING FACTOR VEHICLE 4    1341525
CONTRIBUTING FACTOR VEHICLE 5    1355639
UNIQUE KEY                             0
VEHICLE TYPE COD

**Retriving the Not NULL values in each columns **

In [91]:
print(data.notnull().sum())

DATE                             1360317
TIME                             1360317
BOROUGH                           962675
ZIP CODE                          962551
LATITUDE                         1136013
LONGITUDE                        1136013
LOCATION                         1136013
ON STREET NAME                   1087106
CROSS STREET NAME                 984594
OFF STREET NAME                   224852
NUMBER OF PERSONS INJURED        1360317
NUMBER OF PERSONS KILLED         1360317
NUMBER OF PEDESTRIANS INJURED    1360317
NUMBER OF PEDESTRIANS KILLED     1360317
NUMBER OF CYCLIST INJURED        1360317
NUMBER OF CYCLIST KILLED         1360317
NUMBER OF MOTORIST INJURED       1360317
NUMBER OF MOTORIST KILLED        1360317
CONTRIBUTING FACTOR VEHICLE 1    1353002
CONTRIBUTING FACTOR VEHICLE 2    1164298
CONTRIBUTING FACTOR VEHICLE 3      87647
CONTRIBUTING FACTOR VEHICLE 4      18792
CONTRIBUTING FACTOR VEHICLE 5       4678
UNIQUE KEY                       1360317
VEHICLE TYPE COD

** Here Each columns with less than 1000 records are being dropped
since it wilhave no impact in series due to lack of information.**


In [98]:
for column in data:
    if(data[column].count() < 500000):
        data.drop([column],axis = 1 , inplace=True)

*Now we are left with the columns with less NaN values :)*<br>
**We have reduced columns from 29 to 22**

In [99]:
print(data.notnull().sum())

DATE                             1360317
TIME                             1360317
BOROUGH                           962675
ZIP CODE                          962551
LATITUDE                         1136013
LONGITUDE                        1136013
LOCATION                         1136013
ON STREET NAME                   1087106
CROSS STREET NAME                 984594
NUMBER OF PERSONS INJURED        1360317
NUMBER OF PERSONS KILLED         1360317
NUMBER OF PEDESTRIANS INJURED    1360317
NUMBER OF PEDESTRIANS KILLED     1360317
NUMBER OF CYCLIST INJURED        1360317
NUMBER OF CYCLIST KILLED         1360317
NUMBER OF MOTORIST INJURED       1360317
NUMBER OF MOTORIST KILLED        1360317
CONTRIBUTING FACTOR VEHICLE 1    1353002
CONTRIBUTING FACTOR VEHICLE 2    1164298
UNIQUE KEY                       1360317
VEHICLE TYPE CODE 1              1349456
VEHICLE TYPE CODE 2              1144218
dtype: int64


In [100]:
data

DATE   TIME        BOROUGH ZIP CODE   LATITUDE  LONGITUDE  \
0        10/08/2018   0:00       BROOKLYN    11211  40.713074 -73.952095   
1        10/08/2018   0:00       BROOKLYN    11224  40.577908 -74.008180   
2        10/08/2018   0:00      MANHATTAN    10036  40.758680 -73.983360   
3        10/08/2018   0:00            NaN      NaN  40.708324 -73.843140   
4        10/08/2018   0:00            NaN      NaN  40.763428 -73.965220   
5        10/08/2018   0:00         QUEENS    11435  40.700768 -73.810190   
6        10/08/2018   0:01            NaN      NaN  40.704712 -73.727425   
7        10/08/2018   0:10      MANHATTAN    10013  40.724564 -74.007790   
8        10/08/2018   0:15          BRONX    10469  40.871810 -73.855560   
9        10/08/2018   0:17            NaN      NaN        NaN        NaN   
10       10/08/2018   0:17            NaN      NaN  40.706944 -73.917740   
11       10/08/2018   0:25      MANHATTAN    10006  40.710114 -74.012665   
12       10/08/2018   0:30          BRONX    10472  40.831287 -73.881730   
13       10/08/2018   0:30      MANHATTAN    10027  40.807960 -73.949080   
14       10/08/2018   0:30            NaN      NaN  40.716030 -73.816370   
15       10/08/2018   0:44       BROOKLYN    11201  40.691586 -73.999270   
16       10/08/2018   0:44            NaN      NaN  40.753513 -73.988790   
17       10/08/2018   0:45         QUEENS    11355  40.740173 -73.811264   
18       10/08/2018  10:00            NaN      NaN  40.766434 -73.837670   
19       10/08/2018  10:00            NaN      NaN  40.592846 -73.796160   
20       10/08/2018  10:00            NaN      NaN  40.716957 -73.826550   
21       10/08/2018  10:15            NaN      NaN  40.764587 -73.904520   
22       10/08/2018  10:15            NaN      NaN  40.860600 -73.911020   
23       10/08/2018  10:16      MANHATTAN    10029  40.788580 -73.955284   
24       10/08/2018  10:25          BRONX    10473  40.822086 -73.851906   
25       10/08/2018  10:29            NaN      NaN  40.663315 -73.989770   
26       10/08/2018  10:29            NaN      NaN  40.859850 -73.935520   
27       10/08/2018  10:30       BROOKLYN    11204  40.633500 -73.980520   
28       10/08/2018  10:30         QUEENS    11373  40.740402 -73.887100   
29       10/08/2018  10:30         QUEENS    11418  40.699290 -73.831540   
...             ...    ...            ...      ...        ...        ...   
1360287  07/01/2012   8:27       BROOKLYN    11216  40.680857 -73.938546   
1360288  07/01/2012   8:27            NaN      NaN  40.708797 -73.727620   
1360289  07/01/2012   8:30            NaN      NaN        NaN        NaN   
1360290  07/01/2012   8:34       BROOKLYN    11213  40.668798 -73.931120   
1360291  07/01/2012   8:37         QUEENS    11435  40.694545 -73.802112   
1360292  07/01/2012   8:43            NaN      NaN        NaN        NaN   
1360293  07/01/2012   8:58          BRONX    10466  40.891840 -73.862740   
1360294  07/01/2012   9:00         QUEENS    11101  40.744131 -73.927047   
1360295  07/01/2012   9:00      MANHATTAN    10022  40.757859 -73.975592   
1360296  07/01/2012   9:15       BROOKLYN    11233  40.675380 -73.908196   
1360297  07/01/2012   9:15         QUEENS    11421  40.696308 -73.851549   
1360298  07/01/2012   9:15         QUEENS    11429  40.711860 -73.735538   
1360299  07/01/2012   9:15  STATEN ISLAND    10310  40.634748 -74.112253   
1360300  07/01/2012   9:20            NaN      NaN        NaN        NaN   
1360301  07/01/2012   9:24         QUEENS    11411  40.690917 -73.728179   
1360302  07/01/2012   9:30          BRONX    10475  40.861637 -73.824326   
1360303  07/01/2012   9:30       BROOKLYN    11219  40.626826 -73.996790   
1360304  07/01/2012   9:30         QUEENS    11368  40.755336 -73.843245   
1360305  07/01/2012   9:30       BROOKLYN    11207  40.678163 -73.897477   
1360306  07/01/2012   9:39       BROOKLYN    11226  40.646518 -73.958105   
1360307  07/01/2012   9:40          BRONX    10462  40.839882 -

###  Filling NaN values with 0 for Interger values columns like 
NUMBER OF PERSONS INJURED        
NUMBER OF PERSONS KILLED         
NUMBER OF PEDESTRIANS INJURED    
NUMBER OF PEDESTRIANS KILLED     
NUMBER OF CYCLIST INJURED        
NUMBER OF CYCLIST KILLED         
NUMBER OF MOTORIST INJURED       
NUMBER OF MOTORIST KILLED       

In [101]:
data['NUMBER OF PERSONS INJURED'].fillna(0).head()

0    0
1    0
2    0
3    0
4    0
Name: NUMBER OF PERSONS INJURED, dtype: int64

In [102]:
data['NUMBER OF PERSONS KILLED'].fillna(0).head()

0    0
1    0
2    0
3    0
4    0
Name: NUMBER OF PERSONS KILLED, dtype: int64

In [75]:
df['NUMBER OF PEDESTRIANS INJURED'].fillna(0).head()

0    0
1    0
2    0
3    0
4    0
Name: NUMBER OF PEDESTRIANS INJURED, dtype: int64

In [103]:
data['NUMBER OF PEDESTRIANS KILLED'].fillna(0).head()

0    0
1    0
2    0
3    0
4    0
Name: NUMBER OF PEDESTRIANS KILLED, dtype: int64

In [77]:
df['NUMBER OF CYCLIST INJURED'].fillna(0).head()

0    0
1    0
2    0
3    0
4    0
Name: NUMBER OF CYCLIST INJURED, dtype: int64

In [104]:
data['NUMBER OF CYCLIST KILLED'].fillna(0).head()

0    0
1    0
2    0
3    0
4    0
Name: NUMBER OF CYCLIST KILLED, dtype: int64

In [105]:
data['NUMBER OF MOTORIST INJURED'].fillna(0).head()

0    0
1    0
2    0
3    0
4    0
Name: NUMBER OF MOTORIST INJURED, dtype: int64

In [106]:
data['NUMBER OF MOTORIST KILLED'].fillna(0).tail()

1360312    0
1360313    0
1360314    0
1360315    0
1360316    0
Name: NUMBER OF MOTORIST KILLED, dtype: int64

In [107]:
data

DATE   TIME        BOROUGH ZIP CODE   LATITUDE  LONGITUDE  \
0        10/08/2018   0:00       BROOKLYN    11211  40.713074 -73.952095   
1        10/08/2018   0:00       BROOKLYN    11224  40.577908 -74.008180   
2        10/08/2018   0:00      MANHATTAN    10036  40.758680 -73.983360   
3        10/08/2018   0:00            NaN      NaN  40.708324 -73.843140   
4        10/08/2018   0:00            NaN      NaN  40.763428 -73.965220   
5        10/08/2018   0:00         QUEENS    11435  40.700768 -73.810190   
6        10/08/2018   0:01            NaN      NaN  40.704712 -73.727425   
7        10/08/2018   0:10      MANHATTAN    10013  40.724564 -74.007790   
8        10/08/2018   0:15          BRONX    10469  40.871810 -73.855560   
9        10/08/2018   0:17            NaN      NaN        NaN        NaN   
10       10/08/2018   0:17            NaN      NaN  40.706944 -73.917740   
11       10/08/2018   0:25      MANHATTAN    10006  40.710114 -74.012665   
12       10/08/2018   0:30          BRONX    10472  40.831287 -73.881730   
13       10/08/2018   0:30      MANHATTAN    10027  40.807960 -73.949080   
14       10/08/2018   0:30            NaN      NaN  40.716030 -73.816370   
15       10/08/2018   0:44       BROOKLYN    11201  40.691586 -73.999270   
16       10/08/2018   0:44            NaN      NaN  40.753513 -73.988790   
17       10/08/2018   0:45         QUEENS    11355  40.740173 -73.811264   
18       10/08/2018  10:00            NaN      NaN  40.766434 -73.837670   
19       10/08/2018  10:00            NaN      NaN  40.592846 -73.796160   
20       10/08/2018  10:00            NaN      NaN  40.716957 -73.826550   
21       10/08/2018  10:15            NaN      NaN  40.764587 -73.904520   
22       10/08/2018  10:15            NaN      NaN  40.860600 -73.911020   
23       10/08/2018  10:16      MANHATTAN    10029  40.788580 -73.955284   
24       10/08/2018  10:25          BRONX    10473  40.822086 -73.851906   
25       10/08/2018  10:29            NaN      NaN  40.663315 -73.989770   
26       10/08/2018  10:29            NaN      NaN  40.859850 -73.935520   
27       10/08/2018  10:30       BROOKLYN    11204  40.633500 -73.980520   
28       10/08/2018  10:30         QUEENS    11373  40.740402 -73.887100   
29       10/08/2018  10:30         QUEENS    11418  40.699290 -73.831540   
...             ...    ...            ...      ...        ...        ...   
1360287  07/01/2012   8:27       BROOKLYN    11216  40.680857 -73.938546   
1360288  07/01/2012   8:27            NaN      NaN  40.708797 -73.727620   
1360289  07/01/2012   8:30            NaN      NaN        NaN        NaN   
1360290  07/01/2012   8:34       BROOKLYN    11213  40.668798 -73.931120   
1360291  07/01/2012   8:37         QUEENS    11435  40.694545 -73.802112   
1360292  07/01/2012   8:43            NaN      NaN        NaN        NaN   
1360293  07/01/2012   8:58          BRONX    10466  40.891840 -73.862740   
1360294  07/01/2012   9:00         QUEENS    11101  40.744131 -73.927047   
1360295  07/01/2012   9:00      MANHATTAN    10022  40.757859 -73.975592   
1360296  07/01/2012   9:15       BROOKLYN    11233  40.675380 -73.908196   
1360297  07/01/2012   9:15         QUEENS    11421  40.696308 -73.851549   
1360298  07/01/2012   9:15         QUEENS    11429  40.711860 -73.735538   
1360299  07/01/2012   9:15  STATEN ISLAND    10310  40.634748 -74.112253   
1360300  07/01/2012   9:20            NaN      NaN        NaN        NaN   
1360301  07/01/2012   9:24         QUEENS    11411  40.690917 -73.728179   
1360302  07/01/2012   9:30          BRONX    10475  40.861637 -73.824326   
1360303  07/01/2012   9:30       BROOKLYN    11219  40.626826 -73.996790   
1360304  07/01/2012   9:30         QUEENS    11368  40.755336 -73.843245   
1360305  07/01/2012   9:30       BROOKLYN    11207  40.678163 -73.897477   
1360306  07/01/2012   9:39       BROOKLYN    11226  40.646518 -73.958105   
1360307  07/01/2012   9:40          BRONX    10462  40.839882 -

### Dropping Rows not having atleast 18 real values or having 4 NaN values 
BY  setting Thres value in dropna to 18 means that out of 22 columns there should be atleast 18 columns that should have real values.

In [110]:
data.dropna(thresh=18 ,inplace=True)

In [111]:
data

DATE   TIME        BOROUGH ZIP CODE   LATITUDE  LONGITUDE  \
0        10/08/2018   0:00       BROOKLYN    11211  40.713074 -73.952095   
1        10/08/2018   0:00       BROOKLYN    11224  40.577908 -74.008180   
2        10/08/2018   0:00      MANHATTAN    10036  40.758680 -73.983360   
3        10/08/2018   0:00            NaN      NaN  40.708324 -73.843140   
4        10/08/2018   0:00            NaN      NaN  40.763428 -73.965220   
5        10/08/2018   0:00         QUEENS    11435  40.700768 -73.810190   
6        10/08/2018   0:01            NaN      NaN  40.704712 -73.727425   
7        10/08/2018   0:10      MANHATTAN    10013  40.724564 -74.007790   
8        10/08/2018   0:15          BRONX    10469  40.871810 -73.855560   
10       10/08/2018   0:17            NaN      NaN  40.706944 -73.917740   
11       10/08/2018   0:25      MANHATTAN    10006  40.710114 -74.012665   
12       10/08/2018   0:30          BRONX    10472  40.831287 -73.881730   
13       10/08/2018   0:30      MANHATTAN    10027  40.807960 -73.949080   
14       10/08/2018   0:30            NaN      NaN  40.716030 -73.816370   
15       10/08/2018   0:44       BROOKLYN    11201  40.691586 -73.999270   
16       10/08/2018   0:44            NaN      NaN  40.753513 -73.988790   
17       10/08/2018   0:45         QUEENS    11355  40.740173 -73.811264   
18       10/08/2018  10:00            NaN      NaN  40.766434 -73.837670   
19       10/08/2018  10:00            NaN      NaN  40.592846 -73.796160   
20       10/08/2018  10:00            NaN      NaN  40.716957 -73.826550   
21       10/08/2018  10:15            NaN      NaN  40.764587 -73.904520   
22       10/08/2018  10:15            NaN      NaN  40.860600 -73.911020   
23       10/08/2018  10:16      MANHATTAN    10029  40.788580 -73.955284   
24       10/08/2018  10:25          BRONX    10473  40.822086 -73.851906   
25       10/08/2018  10:29            NaN      NaN  40.663315 -73.989770   
26       10/08/2018  10:29            NaN      NaN  40.859850 -73.935520   
27       10/08/2018  10:30       BROOKLYN    11204  40.633500 -73.980520   
28       10/08/2018  10:30         QUEENS    11373  40.740402 -73.887100   
29       10/08/2018  10:30         QUEENS    11418  40.699290 -73.831540   
30       10/08/2018  10:30            NaN      NaN  40.806133 -73.961555   
...             ...    ...            ...      ...        ...        ...   
1360281  07/01/2012   8:00       BROOKLYN    11233  40.683160 -73.938078   
1360282  07/01/2012   8:00         QUEENS    11004  40.751826 -73.719206   
1360284  07/01/2012   8:07      MANHATTAN    10009  40.721678 -73.980605   
1360285  07/01/2012   8:08       BROOKLYN    11233  40.672791 -73.922215   
1360286  07/01/2012   8:15         QUEENS    11373  40.739979 -73.877108   
1360287  07/01/2012   8:27       BROOKLYN    11216  40.680857 -73.938546   
1360290  07/01/2012   8:34       BROOKLYN    11213  40.668798 -73.931120   
1360291  07/01/2012   8:37         QUEENS    11435  40.694545 -73.802112   
1360293  07/01/2012   8:58          BRONX    10466  40.891840 -73.862740   
1360294  07/01/2012   9:00         QUEENS    11101  40.744131 -73.927047   
1360295  07/01/2012   9:00      MANHATTAN    10022  40.757859 -73.975592   
1360296  07/01/2012   9:15       BROOKLYN    11233  40.675380 -73.908196   
1360297  07/01/2012   9:15         QUEENS    11421  40.696308 -73.851549   
1360298  07/01/2012   9:15         QUEENS    11429  40.711860 -73.735538   
1360299  07/01/2012   9:15  STATEN ISLAND    10310  40.634748 -74.112253   
1360301  07/01/2012   9:24         QUEENS    11411  40.690917 -73.728179   
1360302  07/01/2012   9:30          BRONX    10475  40.861637 -73.824326   
1360303  07/01/2012   9:30       BROOKLYN    11219  40.626826 -73.996790   
1360304  07/01/2012   9:30         QUEENS    11368  40.755336 -73.843245   
1360305  07/01/2012   9:30       BROOKLYN    11207  40.678163 -73.897477   
1360306  07/01/2012   9:39       BROOKLYN    11226  40.646518 -

### *Now we have reduced rows from 1360317 to 1130042* 


## Re-Indexing :
Conform DataFrame to new index with optional filling logic, placing NA/NaN in locations having no value in the previous index. A new object is produced unless the new index is equivalent to the current one and copy=False

In [116]:
data = data.reset_index(drop=True)

In [117]:
data

DATE   TIME        BOROUGH ZIP CODE   LATITUDE  LONGITUDE  \
0        10/08/2018   0:00       BROOKLYN    11211  40.713074 -73.952095   
1        10/08/2018   0:00       BROOKLYN    11224  40.577908 -74.008180   
2        10/08/2018   0:00      MANHATTAN    10036  40.758680 -73.983360   
3        10/08/2018   0:00            NaN      NaN  40.708324 -73.843140   
4        10/08/2018   0:00            NaN      NaN  40.763428 -73.965220   
5        10/08/2018   0:00         QUEENS    11435  40.700768 -73.810190   
6        10/08/2018   0:01            NaN      NaN  40.704712 -73.727425   
7        10/08/2018   0:10      MANHATTAN    10013  40.724564 -74.007790   
8        10/08/2018   0:15          BRONX    10469  40.871810 -73.855560   
9        10/08/2018   0:17            NaN      NaN  40.706944 -73.917740   
10       10/08/2018   0:25      MANHATTAN    10006  40.710114 -74.012665   
11       10/08/2018   0:30          BRONX    10472  40.831287 -73.881730   
12       10/08/2018   0:30      MANHATTAN    10027  40.807960 -73.949080   
13       10/08/2018   0:30            NaN      NaN  40.716030 -73.816370   
14       10/08/2018   0:44       BROOKLYN    11201  40.691586 -73.999270   
15       10/08/2018   0:44            NaN      NaN  40.753513 -73.988790   
16       10/08/2018   0:45         QUEENS    11355  40.740173 -73.811264   
17       10/08/2018  10:00            NaN      NaN  40.766434 -73.837670   
18       10/08/2018  10:00            NaN      NaN  40.592846 -73.796160   
19       10/08/2018  10:00            NaN      NaN  40.716957 -73.826550   
20       10/08/2018  10:15            NaN      NaN  40.764587 -73.904520   
21       10/08/2018  10:15            NaN      NaN  40.860600 -73.911020   
22       10/08/2018  10:16      MANHATTAN    10029  40.788580 -73.955284   
23       10/08/2018  10:25          BRONX    10473  40.822086 -73.851906   
24       10/08/2018  10:29            NaN      NaN  40.663315 -73.989770   
25       10/08/2018  10:29            NaN      NaN  40.859850 -73.935520   
26       10/08/2018  10:30       BROOKLYN    11204  40.633500 -73.980520   
27       10/08/2018  10:30         QUEENS    11373  40.740402 -73.887100   
28       10/08/2018  10:30         QUEENS    11418  40.699290 -73.831540   
29       10/08/2018  10:30            NaN      NaN  40.806133 -73.961555   
...             ...    ...            ...      ...        ...        ...   
1130012  07/01/2012   8:00       BROOKLYN    11233  40.683160 -73.938078   
1130013  07/01/2012   8:00         QUEENS    11004  40.751826 -73.719206   
1130014  07/01/2012   8:07      MANHATTAN    10009  40.721678 -73.980605   
1130015  07/01/2012   8:08       BROOKLYN    11233  40.672791 -73.922215   
1130016  07/01/2012   8:15         QUEENS    11373  40.739979 -73.877108   
1130017  07/01/2012   8:27       BROOKLYN    11216  40.680857 -73.938546   
1130018  07/01/2012   8:34       BROOKLYN    11213  40.668798 -73.931120   
1130019  07/01/2012   8:37         QUEENS    11435  40.694545 -73.802112   
1130020  07/01/2012   8:58          BRONX    10466  40.891840 -73.862740   
1130021  07/01/2012   9:00         QUEENS    11101  40.744131 -73.927047   
1130022  07/01/2012   9:00      MANHATTAN    10022  40.757859 -73.975592   
1130023  07/01/2012   9:15       BROOKLYN    11233  40.675380 -73.908196   
1130024  07/01/2012   9:15         QUEENS    11421  40.696308 -73.851549   
1130025  07/01/2012   9:15         QUEENS    11429  40.711860 -73.735538   
1130026  07/01/2012   9:15  STATEN ISLAND    10310  40.634748 -74.112253   
1130027  07/01/2012   9:24         QUEENS    11411  40.690917 -73.728179   
1130028  07/01/2012   9:30          BRONX    10475  40.861637 -73.824326   
1130029  07/01/2012   9:30       BROOKLYN    11219  40.626826 -73.996790   
1130030  07/01/2012   9:30         QUEENS    11368  40.755336 -73.843245   
1130031  07/01/2012   9:30       BROOKLYN    11207  40.678163 -73.897477   
1130032  07/01/2012   9:39       BROOKLYN    11226  40.646518 -

## Date Parsing :
### In this we parse the date to the format of date as given date is of string format so that it is easy to perform date and time operations. i.e converting from 10/08/2018 to 2018-10-08


In [121]:
data['DATE'] = pd.to_datetime(data['DATE'])

In [122]:
data

DATE   TIME        BOROUGH ZIP CODE   LATITUDE  LONGITUDE  \
0       2018-10-08   0:00       BROOKLYN    11211  40.713074 -73.952095   
1       2018-10-08   0:00       BROOKLYN    11224  40.577908 -74.008180   
2       2018-10-08   0:00      MANHATTAN    10036  40.758680 -73.983360   
3       2018-10-08   0:00            NaN      NaN  40.708324 -73.843140   
4       2018-10-08   0:00            NaN      NaN  40.763428 -73.965220   
5       2018-10-08   0:00         QUEENS    11435  40.700768 -73.810190   
6       2018-10-08   0:01            NaN      NaN  40.704712 -73.727425   
7       2018-10-08   0:10      MANHATTAN    10013  40.724564 -74.007790   
8       2018-10-08   0:15          BRONX    10469  40.871810 -73.855560   
9       2018-10-08   0:17            NaN      NaN  40.706944 -73.917740   
10      2018-10-08   0:25      MANHATTAN    10006  40.710114 -74.012665   
11      2018-10-08   0:30          BRONX    10472  40.831287 -73.881730   
12      2018-10-08   0:30      MANHATTAN    10027  40.807960 -73.949080   
13      2018-10-08   0:30            NaN      NaN  40.716030 -73.816370   
14      2018-10-08   0:44       BROOKLYN    11201  40.691586 -73.999270   
15      2018-10-08   0:44            NaN      NaN  40.753513 -73.988790   
16      2018-10-08   0:45         QUEENS    11355  40.740173 -73.811264   
17      2018-10-08  10:00            NaN      NaN  40.766434 -73.837670   
18      2018-10-08  10:00            NaN      NaN  40.592846 -73.796160   
19      2018-10-08  10:00            NaN      NaN  40.716957 -73.826550   
20      2018-10-08  10:15            NaN      NaN  40.764587 -73.904520   
21      2018-10-08  10:15            NaN      NaN  40.860600 -73.911020   
22      2018-10-08  10:16      MANHATTAN    10029  40.788580 -73.955284   
23      2018-10-08  10:25          BRONX    10473  40.822086 -73.851906   
24      2018-10-08  10:29            NaN      NaN  40.663315 -73.989770   
25      2018-10-08  10:29            NaN      NaN  40.859850 -73.935520   
26      2018-10-08  10:30       BROOKLYN    11204  40.633500 -73.980520   
27      2018-10-08  10:30         QUEENS    11373  40.740402 -73.887100   
28      2018-10-08  10:30         QUEENS    11418  40.699290 -73.831540   
29      2018-10-08  10:30            NaN      NaN  40.806133 -73.961555   
...            ...    ...            ...      ...        ...        ...   
1130012 2012-07-01   8:00       BROOKLYN    11233  40.683160 -73.938078   
1130013 2012-07-01   8:00         QUEENS    11004  40.751826 -73.719206   
1130014 2012-07-01   8:07      MANHATTAN    10009  40.721678 -73.980605   
1130015 2012-07-01   8:08       BROOKLYN    11233  40.672791 -73.922215   
1130016 2012-07-01   8:15         QUEENS    11373  40.739979 -73.877108   
1130017 2012-07-01   8:27       BROOKLYN    11216  40.680857 -73.938546   
1130018 2012-07-01   8:34       BROOKLYN    11213  40.668798 -73.931120   
1130019 2012-07-01   8:37         QUEENS    11435  40.694545 -73.802112   
1130020 2012-07-01   8:58          BRONX    10466  40.891840 -73.862740   
1130021 2012-07-01   9:00         QUEENS    11101  40.744131 -73.927047   
1130022 2012-07-01   9:00      MANHATTAN    10022  40.757859 -73.975592   
1130023 2012-07-01   9:15       BROOKLYN    11233  40.675380 -73.908196   
1130024 2012-07-01   9:15         QUEENS    11421  40.696308 -73.851549   
1130025 2012-07-01   9:15         QUEENS    11429  40.711860 -73.735538   
1130026 2012-07-01   9:15  STATEN ISLAND    10310  40.634748 -74.112253   
1130027 2012-07-01   9:24         QUEENS    11411  40.690917 -73.728179   
1130028 2012-07-01   9:30          BRONX    10475  40.861637 -73.824326   
1130029 2012-07-01   9:30       BROOKLYN    11219  40.626826 -73.996790   
1130030 2012-07-01   9:30         QUEENS    11368  40.755336 -73.843245   
1130031 2012-07-01   9:30       BROOKLYN    11207  40.678163 -73.897477   
1130032 2012-07-01   9:39       BROOKLYN    11226  40.646518 -73.958105   
1130033 2012-07-01   9:40          BRON